exp/bufferpool/pool.go

```diff
@@ -0,0 +1,53 @@
+// Package bufferpool supports object pooling for byte buffers.
+package bufferpool
+
+import "sync"
+
+const (
+	nBuckets = 20
+)
+
+// A default global pool.
+var Default Pool
+
+// A pool of buffers, in variable sizes.
+type Pool struct {
+	buckets [nBuckets]sync.Pool
+}
+
+// Get a buffer of the given length. its capacity may be larger than the
+// requested size.
+func (p *Pool) Get(size int) []byte {
+
+	for i := 0; i < nBuckets; i++ {
+		capacity := 1 << i
+		if capacity >= size {
+			var ret []byte
+			item := p.buckets[i].Get()
+			if item == nil {
+				ret = make([]byte, capacity)
+			} else {
+				ret = item.([]byte)
+			}
+			ret = ret[:size]
+			return ret
+		}
+	}
+	return make([]byte, size)
+}
+
+// Return a buffer to the pool. Zeros the slice (but not the full backing array)
+// before making it available for future use.
+func (p *Pool) Put(buf []byte) {
+	for i := 0; i < len(buf); i++ {
+		buf[i] = 0
+	}
+
+	capacity := cap(buf)
```

Comment:
Not 100% sure, but I think the compiler is able to optimize away an inline call to `cap(buf)` due to the capacity of the backing array never changing.  You might be able to write `if (1 << i) == cap(buf) { ... }`.

I don't know what the performance implications are of assigning a new variable.  Probably negligible.

(I'm partly bringing this up because you know a lot about this stuff, and I'm curious.)

Reply:
There wasn't really any performance consideration here; I suspect you're right and the compiler is smart enough to handle that.  I just found it easier to follow when I could mentally treat that as a unit.

Reply:
Yeah, that's what I figured.  Ok, carry on 🙂 👍